reprex 2.0.0 (#507)
tidyverse/tidyverse.org: content/blog/reprex-2-0-0/index.Rmd

---
output: hugodown::hugo_document

slug: reprex-2-0-0
title: reprex 2.0.0
date: 2021-04-02
author: Jenny Bryan
description: >
    reprex is now much easier to use on RStudio Server and RStudio Cloud.
photo:
  url: https://unsplash.com/photos/NROHA1B-NYk
  author: Mitchell Luo
categories: [package] 
---

We're overjoyed to announce the release of [reprex](https://reprex.tidyverse.org) 2.0.0.
reprex is a package that helps you prepare **REPR**oducible **EX**amples to share in places where people talk about code, e.g., on GitHub, on Stack Overflow, and in Slack or email messages.

You can install the current version of reprex from CRAN with[^1]:

[^1]: Another way you might get reprex is by installing the tidyverse meta-package.

> reprex is one of the packages installed by `install.packages("tidyverse")`, however it is **not** among the core packages attached by `library(tidyverse)`.

Reviewer: ```suggestion
reprex is one of the packages installed by `install.packages("tidyverse")`, however it is **not** among the [core packages](https://www.tidyverse.org/packages/#core-tidyverse) attached by `library(tidyverse)`.
```

Reviewer: Feel free to reject, just thought it might be nice, since it sort of begs the question, _Which packages are core?_

```{r, eval = FALSE}
install.packages("reprex")
```

> reprex recently had a major release ([version 1.0.0](https://www.tidyverse.org/blog/2021/02/reprex-1-0-0/)), but we've added some big features since then.

Reviewer: ```suggestion
reprex recently had a major release, ([version 1.0.0](https://www.tidyverse.org/blog/2021/02/reprex-1-0-0/)), but we've added some big features since then.
```

> Specifically, reprex has gotten much, much easier to use on [RStudio Server](https://www.rstudio.com/products/rstudio/download-server/) and on [RStudio Cloud](https://rstudio.cloud).

Reviewer: ```suggestion
Specifically, reprex has gotten much, much easier to use on [RStudio Server](https://www.rstudio.com/products/rstudio/download-server/) and [RStudio Cloud](https://rstudio.cloud).
```

Reviewer: Nixed it, since it's the same preposition for both.

It's also easier to specify the working directory, if you must, and reprex plays more nicely with the [renv package](https://rstudio.github.io/renv/).

You can see a full list of changes in the [release notes](https://reprex.tidyverse.org/news/index.html#reprex-2-0-0-2021-04-02).

```{r setup}
library(reprex)
```

## When the clipboard isn't available

By default, reprex accepts a code snippet via the clipboard and also puts the rendered result there, ready to paste into GitHub, Stack Overflow, Slack or an email.
But when working in [RStudio Server](https://www.rstudio.com/products/rstudio/download-server/) (RS) or [RStudio Cloud](https://rstudio.cloud), you're running R (and reprex) in a web browser.

For very good security reasons, it's essentially impossible to access your clipboard from R in this context.
Therefore, reprex needs to use other methods to remove as much friction as possible, both for input and output.

> When `reprex()` is called without any code input, in RS or on Cloud, the default is now to **consult the current selection** for reprex source.

Reviewer 2: ```suggestion
When `reprex()` is called without any code input, in RStudio Server or on Cloud, the default is now to **consult the current selection** for reprex source.
```

Previously this was only available via the `reprex_selection()` addin. [^2]

> Note that this "current selection" default behaviour propagates to convenience wrappers around `reprex()`, such as `reprex_locale()` and venue-specific functions `reprex_r()`, as well as to the un-`reprex()` functions, such as `reprex_clean()`.

Reviewer 2: I think this could go in a footnote

[^2]: The `reprex_selection()` addin is alive and well!
What's new is that it has become the default behaviour on RS/Cloud.

> `reprex_selection()` is so handy that many reprex users wire it up to a [custom keyboard shortcut](https://support.rstudio.com/hc/en-us/articles/206382178-Customizing-Keyboard-Shortcuts), such as Cmd + Shift + R (macOS) or Ctrl + Shift + R (Windows).

Reviewer: I'll open a separate issue for this, but I wonder if it would be worth adding a bit of CSS so we could use `<kbd>` markup for keyboard shortcuts, etc. (I just tried adding them in an extant post, and it doesn't do anything.)

Finally, in RS or on Cloud, the file containing the rendered reprex is opened and its contents are selected, ready for the user to copy via Cmd/Ctrl + C.

These changes also make the ["Render reprex" addin/gadget](https://reprex.tidyverse.org/reference/reprex_addin.html) much more usable on RS and on Cloud. 

## Working directory

> Ideally, a reprex is entirely self-contained and does not read any local files that, presumably, nobody else has.

Reviewer: ```suggestion
Ideally, a reprex is entirely self-contained and does not read any local files which, presumably, nobody else has.
```

Reviewer: 🤓  Me being fidgety because the files are the DO of "nobody else has."

That's why, by default, `reprex()` works in an ephemeral directory created in the session temp directory.
Run `getwd()` inside `reprex()`:

```{r, eval = FALSE}
reprex::reprex(
  getwd()
)
```

and you'll see something like this:

``` r
getwd()
#> [1] "/private/tmp/RtmpXrUKXE/reprex-13063117cb6d8-alive-lark"
```

However, in the real world, sometimes you must read a local data file that you can't share with the world.
You can still at least use `reprex()` to reveal your actual code and results.

This has always been possible, but we've reworked some arguments to make this more natural to express:

* `wd` is a new argument to set the reprex working directory
* `outfile` is deprecated, with the existing `input` argument taking over its
  duties

If we run:

```{r, eval = FALSE}
reprex::reprex(
  getwd(),
  wd = "."
)
```

we see the working directory is the directory where the source of this blog post lives:

``` r
getwd()
#> [1] "/Users/jenny/rrr/tidyverse.org/content/blog/reprex-2-0-0"
```

The old way to `reprex()` in current working directory was `reprex(outfile =  NA)`, which was not very intuitive.

> You're not limited to `wd = "."`; you can use `wd` with any existing path.

Reviewer 2: I think this can go without saying. I know not everyone will realise it, but I think the people who don't understand are most likely to use the `wd` appropriately.

### Adjective-animal

Various changes mean that more users will see reprex filepaths.
Therefore, we've revised them to be more self-explanatory and human-friendly.
When reprex needs to invent a file name, it is now based on a random "adjective-animal" slug.
Bring on the `angry-hamster`! [^3]

[^3]: `reprex()` draws randomly from a fixed set of adjective-animal slugs generated with the handy [ids package](https://reside-ic.github.io/ids/).

You actually saw one of these filepaths already, in the output of `reprex(getwd())`:

``` r
getwd()
#> [1] "/private/tmp/RtmpXrUKXE/reprex-13063117cb6d8-alive-lark"
```

> ## Local `.Rprofile` and renv happiness

Reviewer 2: I think this section should follow the sectoin about `wd`, and be explained in that context.
